builder.go

```diff
@@ -175,25 +206,54 @@ func containsEntry(metRequires []string, entry BuildPlanEntry) bool {
 	return false
 }
 
-type processMap map[string]launch.Process
+// orderedProcesses is a mapping from process types to Processes, it will preserve ordering.
+// processList is the ordered list.
+// we keep typeToProcess map in order to delete elements when others are overriding them.
+type orderedProcesses struct {
+	typeToProcess map[string]*list.Element
+	processList   *list.List
+}
 
-func (m processMap) add(l []launch.Process) {
-	for _, proc := range l {
-		m[proc.Type] = proc
+func newProcessStruct() orderedProcesses {
+	return orderedProcesses{
+		typeToProcess: make(map[string]*list.Element),
+		processList:   list.New(),
 	}
 }
 
-func (m processMap) list() []launch.Process {
-	var keys []string
-	for key := range m {
-		keys = append(keys, key)
+func (m orderedProcesses) add(listToAdd []launch.Process) ([]string, error) {
+	result := []string{}
+	for _, proc := range listToAdd {
+		// when we replace a default process type with a non-default process type, add to result
+		if p, ok := m.typeToProcess[proc.Type]; ok {
+			cast, success := (p.Value).(launch.Process)
```

Comment:
```suggestion
			existingProc, success := (p.Value).(launch.Process)
```

or `foundProc` maybe

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/520/commits/63f4e9a470ff92f133caa5ffff56d047d6fa3295.